src/boot/efi/random-seed.c

```diff
@@ -221,31 +117,76 @@ static void validate_sha256(void) {
 }
 
 EFI_STATUS process_random_seed(EFI_FILE *root_dir, RandomSeedMode mode) {
-        _cleanup_free_ void *seed = NULL, *new_seed = NULL, *rng = NULL, *for_kernel = NULL, *system_token = NULL;
+        _cleanup_erase_ unsigned char random_bytes[DESIRED_SEED_SIZE], hash_key[HASH_VALUE_SIZE];
+        struct linux_efi_random_seed *prior_seed_table = NULL, *new_seed_table;
+        _cleanup_free_ void *seed = NULL, *system_token = NULL;
         _cleanup_(file_closep) EFI_FILE *handle = NULL;
-        UINTN size, rsize, wsize, system_token_size = 0;
         _cleanup_free_ EFI_FILE_INFO *info = NULL;
+        _cleanup_erase_ struct sha256_ctx hash;
         uint64_t uefi_monotonic_counter = 0;
+        UINTN size, rsize, wsize;
+        unsigned char usage_idx;
+        bool got_rng_bytes;
         EFI_STATUS err;
 
         assert(root_dir);
+        assert(DESIRED_SEED_SIZE == HASH_VALUE_SIZE);
 
         validate_sha256();
 
         if (mode == RANDOM_SEED_OFF)
                 return EFI_NOT_FOUND;
 
-        /* Let's better be safe than sorry, and for now disable this logic in SecureBoot mode, so that we
-         * don't credit a random seed that is not authenticated. */
-        if (secure_boot_enabled())
-                return EFI_NOT_FOUND;
+        /* hash = LABEL || sizeof(input1) || input1 || ... || sizeof(inputN) || inputN || uefi_monotonic */
+        sha256_init_ctx(&hash);
+
+        /* Some basic domain separation in case somebody uses this data elsewhere */
+        sha256_process_bytes(HASH_LABEL, sizeof(HASH_LABEL) - 1, &hash);
+
+        err = LibGetSystemConfigurationTable(LINUX_EFI_RANDOM_SEED_TABLE_GUID, (void **) &prior_seed_table);
```

Comment:
Okay, will do.

Reply:
Done. (Will push shortly.)

Reply:
I'm curious, by the way, why get rid of gnu-efi? I don't know much about it one way or another, but would be interesting to know your reasoning in case I ever make a little EFI thing of my own.

Reply:
Here is how you create a EFI binary with gnu-efi: You compile your code like it is regular userspace code (using sysv calling abi) into elf obj files. And then you magically (I still have no idea how this works) link/glue this together with a PE stub coming from gnu-efi. Because of this, the PE binary has no PE relocations (that would be dealt with by the firmware). Instead, the PE stub contains a ELF(!) relocator that performs life-surgery at runtime. Just describing this makes me want to go have a shower.

Meanwhile, all you need is to compile code like a native windows PE binary (using a windows --target with clang or with mingw) and set the subsystem to efi app. Much cleaner to build and you get stuff like nx (should there ever come a firmware that cares for it) for free.

There is also a general code quality issue with gnu-efi (upstream is basically dead / on life support).

Really, anyone wanting to start something new targeting EFI should just use uefi-rs. Heck, I've been toying with the idea of writing a sd-stub/boot replacement in it for some time but I am lazy (though, I like to think that I am just waiting for error_in_core to get stabilized :cat:).

Reply:
> Instead, the PE stub contains a ELF(!) relocator that performs life-surgery at runtime. 

Hah wow that is insane... Thanks for the explanation. I wonder how calls to EFI runtime functions work - if the gnu-efi runtime replaces all the pointers with trampolines or something.

Reply:
It has two modes. Originally it used a hand-crafted trampolines in assembler that you have to call with that ugly `uefi_call_wrapper` macro. But at some point gcc learned the `ms_abi` function attribute, which all EFI functions must be marked with. Then the macro is just a regular function call leaving this to the compiler to figure out. We enforce this mode now so that we don't have to deal with this crap macro.

Reply:
Oh, that's a good idea.